File: apps/ios/Exponential/UI/MyWork/ReviewsViewModel.swift
import ExpUI
import ExpCore
import Foundation
import GRDB

/// One review entry (EXP-131): the open PR(s) awaiting review. A batch coding
/// run links several issues to ONE `prUrl`, so those issues collapse into a
/// single entry; an issue with no `prUrl` (shouldn't normally happen for an
/// open PR, but be defensive) keys on its own id so it still renders once.
struct ReviewEntry: Identifiable {
    /// `prUrl` when present, else `issue:<id>` — the grouping key.
    let id: String
    /// The issues sharing this PR, newest first. `representative` is the first.
    let issues: [IssueEntity]

    var representative: IssueEntity { issues[0] }
    var isBatch: Bool { issues.count > 1 }
    var prUrl: String? { representative.prUrl }
    var prNumber: Int? { representative.prNumber }
    var branch: String? { representative.branch }
    /// EXP-897: the branch this PR is BASED on — the stack edge.
    var prBaseBranch: String? { representative.prBaseBranch }
    /// Identifiers of every linked issue, newest first — mirrors `issues`
    /// (for the batch row subtitle).
    var identifiers: [String] { issues.compactMap { $0.identifier } }
}

/// EXP-897: one rendered Reviews row — an entry plus its place in its STACK.
/// The list is flat; the depth is what indents it and the caption is what
/// names the pull request it is built on.
struct ReviewRow: Identifiable {
    let entry: ReviewEntry
    /// 0 for a root (a PR based on something nobody here owns), +1 per rung.
    let depth: Int
    /// Something is stacked on this row.
    let hasChildren: Bool
    /// The identifier of the entry directly BELOW — `on top of #EXP-11`.
    let stackedOn: String?
    /// How many pull requests the stack this row roots holds (1 = not a
    /// stack). Only meaningful on the bottom row — the "Merge stack" copy.
    var stackSize: Int = 1
    var id: String { entry.id }

    /// The LOWEST row of a real stack: the one that offers "Merge stack".
    /// Its own issue id is what `mergePr({mergeStack: true})` takes — the
    /// server resolves the top of the chain from it.
    var isStackBottom: Bool { depth == 0 && hasChildren }
}

/// EXP-734: one AGENT RUN's own open pull request — the chore PR an action or
/// chat run opened through `exponential_pr_open({repositoryId, head})`, which
/// links no issue at all and so appears in no board group.
struct RunReviewEntry: Identifiable {
    let session: CodingSessionEntity
    /// The run's own name: its action, or "Chat" for a chat run.
    let title: String
    var id: String { session.id }
    var prUrl: String? { session.prUrl }
    var prNumber: Int? { session.prNumber }
    var branch: String? { session.branch }
}

/// One board's review entries — Reviews groups by board like the other
/// cross-board lists group by status.
struct ReviewGroup: Identifiable {
    let board: BoardEntity
    /// EXP-897: the board's rows in NESTED order — a stack member follows the
    /// pull request it is based on, one level deeper, and a whole stack lives
    /// in the board of its ROOT entry (a stack spanning two boards reads as
    /// one thing, where it starts).
    let rows: [ReviewRow]
    var id: String { board.id }
    /// The flat entries behind the rows — counts and pickers.
    var entries: [ReviewEntry] { rows.map(\.entry) }
}

/// "Reviews" (EXP-131): every issue in the ACTIVE team with an open PR,
/// collapsed to one entry per distinct PR (a batch PR appears once, not N
/// times), grouped by board. Mirrors `MyIssuesViewModel`'s GRDB observation
/// pattern — two independent, cancellable loops over issues + boards.
@MainActor @Observable
final class ReviewsViewModel {
    var issues: [IssueEntity] = []
    var boards: [BoardEntity] = []
    /// EXP-734: issue-less runs whose OWN pull request is open — the chore PRs
    /// no board group can ever show.
    var runSessions: [CodingSessionEntity] = []

    private let accountId: String
    private let db: DatabaseManager

    private var issueTask: Task<Void, Never>?
    private var boardTask: Task<Void, Never>?
    private var sessionTask: Task<Void, Never>?

    init(accountId: String, db: DatabaseManager) {
        self.accountId = accountId
        self.db = db
    }

    func startObserving() {
        stopObserving() // restartable: the view re-arms on every appear
        guard let pool = try? db.pool(forAccountId: accountId) else { return }

        // Only issues with an OPEN PR are review candidates.
        let issueObservation = ValueObservation.tracking { db in
            try IssueEntity
                .filter(Column("pr_state") == DomainContract.prStateOpen)
                .fetchAll(db)
        }
        issueTask = Task { [weak self] in
            do {
                for try await issues in issueObservation.values(in: pool) {
                    self?.issues = issues
                }
            } catch {}
        }

        // Boards resolve each entry's board (name/section) and scope the
        // list to the active team (issues carry no team_id).
        let boardObservation = ValueObservation.tracking { db in
            try BoardEntity.fetchAll(db)
        }
        boardTask = Task { [weak self] in
            do {
                for try await boards in boardObservation.values(in: pool) {
                    self?.boards = boards
                }
            } catch {}
        }

        // EXP-734: an action or chat run's own PR links no issue, so it can
        // only be found on the session row the server stamped it on.
        let sessionObservation = ValueObservation.tracking { db in
            try CodingSessionEntity
                .filter(Column("issue_id") == nil)
                .filter(Column("pr_state") == DomainContract.prStateOpen)
                .fetchAll(db)
        }
        sessionTask = Task { [weak self] in
            do {
                for try await sessions in sessionObservation.values(in: pool) {
                    self?.runSessions = sessions
                }
            } catch {}
        }
    }

    func stopObserving() {
        issueTask?.cancel()
        issueTask = nil
        boardTask?.cancel()
        boardTask = nil
        sessionTask?.cancel()
        sessionTask = nil
    }

    /// Review entries grouped by board, scoped to `teamId`. Entries
    /// within a board are newest-first; board sections follow the sidebar's
    /// `sortOrder`. Empty when no team is active.
    func groups(teamId: String?) -> [ReviewGroup] {
        guard let teamId else { return [] }

        let teamBoards = boards.filter { $0.teamId == teamId }
        let boardById = Dictionary(uniqueKeysWithValues: teamBoards.map { ($0.id, $0) })
        let candidates = issues.filter { boardById[$0.boardId] != nil }

        // Collapse issues sharing a prUrl into one entry (fall back to the issue
        // id when prUrl is absent). Preserve first-seen order for determinism.
        var buckets: [String: [IssueEntity]] = [:]
        var keyOrder: [String] = []
        for issue in candidates {
            let key = (issue.prUrl?.isEmpty == false) ? issue.prUrl! : "issue:\(issue.id)"
            if buckets[key] == nil { keyOrder.append(key); buckets[key] = [] }
            buckets[key]?.append(issue)
        }

        let entries: [ReviewEntry] = keyOrder.compactMap { key in
            guard let bucket = buckets[key], !bucket.isEmpty else { return nil }
            // Newest first inside the entry — representative is the newest issue.
            let sorted = bucket.sorted { Self.newerFirst($0, $1) }
            return ReviewEntry(id: key, issues: sorted)
        }

        // EXP-897: order FIRST, then nest — `nestPrStacks` keeps the caller's
        // root order and threads every stacked entry under the one it is based
        // on, so the newest-first rule survives the nesting.
        let ordered = entries.sorted { Self.newerFirst($0.representative, $1.representative) }
        let nested = PrStack.nestPrStacks(
            ordered,
            id: { $0.id },
            branch: { $0.branch },
            base: { $0.prBaseBranch }
        )

        // Walk the nested list once: a row's board is its ROOT's board, and
        // the entry directly below it is the nearest preceding row one level
        // shallower (`ancestors`).
        var rows: [ReviewRow] = []
        var boardOfRow: [String] = []
        var ancestors: [ReviewEntry] = []
        var rootBoardId = ""
        for nestedRow in nested {
            let entry = nestedRow.entry
            if ancestors.count > nestedRow.depth {
                ancestors.removeSubrange(nestedRow.depth...)
            }
            let below = nestedRow.depth > 0 ? ancestors.last : nil
            if nestedRow.depth == 0 { rootBoardId = entry.representative.boardId }
            rows.append(ReviewRow(
                entry: entry,
                depth: nestedRow.depth,
                hasChildren: nestedRow.hasChildren,
                stackedOn: below?.representative.identifier
            ))
            boardOfRow.append(rootBoardId)
            ancestors.append(entry)
        }

        // The bottom row of a stack carries its size: the contiguous run of
        // deeper rows that follows it.
        for index in rows.indices where rows[index].depth == 0 {
            var size = 1
            var cursor = index + 1
            while cursor < rows.count, rows[cursor].depth > 0 {
                size += 1
                cursor += 1
            }
            rows[index].stackSize = size
        }

        var byBoard: [String: [ReviewRow]] = [:]
        for (index, row) in rows.enumerated() {
            byBoard[boardOfRow[index], default: []].append(row)
        }

        return teamBoards
            .sorted { ($0.sortOrder ?? 0) < ($1.sortOrder ?? 0) }
            .compactMap { board in
                guard let boardRows = byBoard[board.id], !boardRows.isEmpty else { return nil }
                return ReviewGroup(board: board, rows: boardRows)
            }
    }

    /// EXP-734: the team's agent runs parking their OWN open pull request, one
    /// entry per distinct prUrl (newest run wins), newest first. Sessions carry
    /// `team_id`, so no board scope is needed.
    func runEntries(teamId: String?) -> [RunReviewEntry] {
        guard let teamId else { return [] }
        var byPrUrl: [String: CodingSessionEntity] = [:]
        for session in runSessions where session.teamId == teamId {
            guard session.hasOpenPr, let prUrl = session.prUrl else { continue }
            if let current = byPrUrl[prUrl], Self.newerFirst(current, session) { continue }
            byPrUrl[prUrl] = session
        }
        return byPrUrl.values
            .sorted { Self.newerFirst($0, $1) }
            .map {
                RunReviewEntry(
                    session: $0,
                    title: PastRuns.chatSubject($0) ?? $0.actionName ?? PastRuns.chatRunName
                )
            }
    }

    /// Newest-first by `startedAt`, id as the deterministic tie-break — the
    /// issue rule applied to session rows.
    private static func newerFirst(_ a: CodingSessionEntity, _ b: CodingSessionEntity) -> Bool {
        if a.startedAt != b.startedAt { return a.startedAt > b.startedAt }
        return a.id > b.id
    }

    /// Newest-first by `createdAt` (Postgres wire text compares chronologically,
    /// the IssueSorting precedent), id as the deterministic tie-break.
    private static func newerFirst(_ a: IssueEntity, _ b: IssueEntity) -> Bool {
        if a.createdAt != b.createdAt { return a.createdAt > b.createdAt }
        return a.id > b.id
    }
}

File: apps/android/app/src/main/java/com/exponential/app/ui/reviews/ReviewsViewModel.kt
package com.exponential.app.ui.reviews

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.exponential.app.data.TeamSelection
import com.exponential.app.data.api.CodingSessionsApi
import com.exponential.app.data.api.IssuesApi
import com.exponential.app.data.api.WorkflowsApi
import com.exponential.app.data.auth.AuthRepository
import com.exponential.app.data.db.CodingSessionEntity
import com.exponential.app.data.db.DatabaseHolder
import com.exponential.app.data.db.IssueEntity
import com.exponential.app.data.db.BoardEntity
import com.exponential.app.data.db.WorkflowEntity
import com.exponential.app.data.db.WorkflowNodeEntity
import com.exponential.app.data.db.accountDatabaseFlow
import com.exponential.app.domain.CHAT_RUN_NAME
import com.exponential.app.domain.DomainContract
import com.exponential.app.domain.MergeFailure
import com.exponential.app.domain.PrStack
import com.exponential.app.domain.ReviewMergeInput
import com.exponential.app.domain.ReviewStackPosition
import com.exponential.app.domain.ReviewsMerge
import com.exponential.app.domain.coveredIssueIds
import com.exponential.app.domain.WorkflowFinalPr
import com.exponential.app.domain.chatRunSubject
import com.exponential.app.domain.sortableTimestamp
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

// Reviews (EXP-131): every open pull request in the CURRENT team, grouped
// by board. A batch coding run links N issues to ONE pr_url, so the list
// collapses those rows into a single entry (never N). Pure client work over the
// already-synced issues shape — no new shape, no server round-trip to list.

/**
 * One reviewable pull request. A single-issue PR carries one issue; a batch PR
 * carries several (all sharing [prUrl]). [issues] is newest-first, so
 * [representative] is the newest issue — the one merge/navigation acts on.
 */
data class ReviewEntry(
    val groupKey: String,
    val prUrl: String?,
    val prNumber: Int?,
    val branch: String?,
    val boardId: String,
    val issues: List<IssueEntity>,
) {
    val representative: IssueEntity get() = issues.first()
    val isBatch: Boolean get() = issues.size > 1
    val identifiers: List<String> get() = issues.map { it.identifier }
}

/**
 * EXP-734: one reviewable pull request that belongs to a RUN, not an issue —
 * an action or chat run that opened a chore PR via
 * `exponential_pr_open({repositoryId, head})`. Nothing links it to a board, so
 * these list under their own header and merge through
 * `codingSessions.mergePr`.
 */
data class RunReviewEntry(
    val groupKey: String,
    val session: CodingSessionEntity,
    val prUrl: String?,
    val prNumber: Int?,
    val branch: String?,
    /** The run's own name — an action run's snapshot, or plain "Chat". */
    val title: String,
)

/**
 * The open-PR runs → review entries: collapsed by `pr_url` (a resumed run
 * continues the same PR) keeping the NEWEST row, newest first. Top-level and
 * pure so it can be tested without a database.
 */
fun buildRunEntries(sessions: List<CodingSessionEntity>): List<RunReviewEntry> {
    val byPrUrl = LinkedHashMap<String, CodingSessionEntity>()
    for (session in sessions) {
        val prUrl = session.prUrl
        if (prUrl.isNullOrEmpty()) continue
        val current = byPrUrl[prUrl]
        if (current == null ||
            sortableTimestamp(session.startedAt) > sortableTimestamp(current.startedAt)
        ) {
            byPrUrl[prUrl] = session
        }
    }
    return byPrUrl.values
        .sortedByDescending { sortableTimestamp(it.startedAt) }
        .map { session ->
            RunReviewEntry(
                groupKey = "session:${session.id}",
                session = session,
                prUrl = session.prUrl,
                prNumber = session.prNumber,
                branch = session.branch,
                title = chatRunSubject(session) ?: session.actionName ?: CHAT_RUN_NAME,
            )
        }
}

/**
 * EXP-1072: a workflow's ONE final pull request (integration branch → the
 * default branch). The workflow row carries its url/number/state, so it is
 * the workflow's OWN PR here — never an unlinked one — and merges through
 * `workflows.mergeFinalPr`, which completes the workflow and its issues.
 */
data class WorkflowReviewEntry(
    val groupKey: String,
    val workflow: WorkflowEntity,
    val prUrl: String?,
    val prNumber: Int?,
    val branch: String?,
    /** The workflow's name — the row's title. */
    val title: String,
)

/**
 * The team's workflows → review entries: only an OPEN final pull request with
 * a url, newest workflow first. Pure so it can be tested without a database.
 */
fun buildWorkflowEntries(workflows: List<WorkflowEntity>): List<WorkflowReviewEntry> =
    workflows
        .filter { it.finalPrState == DomainContract.prStateOpen && !it.finalPrUrl.isNullOrEmpty() }
        .sortedByDescending { sortableTimestamp(it.createdAt) }
        .map { workflow ->
            WorkflowReviewEntry(
                groupKey = WorkflowFinalPr.reviewKey(workflow.id),
                workflow = workflow,
                prUrl = workflow.finalPrUrl,
                prNumber = workflow.finalPrNumber,
                branch = workflow.integrationBranch.takeIf { it.isNotBlank() },
                title = workflow.name,
            )
        }

/**
 * EXP-897: one row of the Reviews list — a pull request and where it sits in
 * its STACK. The stack edge is synced (`pr_base_branch` → the lower entry's
 * `branch`), so the nesting is pure client work like the batch collapsing
 * above it.
 */
data class ReviewRowEntry(
    val entry: ReviewEntry,
    /** 0 for the bottom of a stack (or a lone PR), +1 per level. */
    val depth: Int,
    /** Whether a stacked pull request is nested right below this row. */
    val hasChildren: Boolean,
    /** The identifier of the entry directly below — the `on top of #X` caption. */
    val stackedOn: String?,
    /** The board the ROOT of this row's stack belongs to — the group it lands in. */
    val rootBoardId: String,
    /**
     * Non-null on the BOTTOM row of a real stack: the issue id `Merge stack`
     * sends (the server resolves the chain's top from it) and how many pull
     * requests that merge would take.
     */
    val mergeStackIssueId: String?,
    val stackSize: Int,
    /**
     * EXP-1094: the status of the workflow whose node covers this PR's
     * issues (a live one first), else null. A running or paused workflow
     * merges its node PRs itself.
     */
    val workflowStatus: String? = null,
) {
    /** EXP-1094: the input of the ONE merge control this row carries. */
    val mergeInput: ReviewMergeInput
        get() = ReviewMergeInput(
            stack = when {
                mergeStackIssueId != null -> ReviewStackPosition.BOTTOM
                depth > 0 -> ReviewStackPosition.UPPER
                else -> ReviewStackPosition.NONE
            },
            workflowStatus = workflowStatus,
        )
}

/**
 * The team's review entries → rows, nested by stack: the caller's order is
 * the ROOT order, a stacked pull request follows its foundation, and every
 * row records the board of its ROOT so a whole stack groups under one board
 * even when a member was moved.
 */
fun buildReviewRows(
    entries: List<ReviewEntry>,
    /** EXP-1094: issue id → its covering workflow's status ([ReviewsMerge.workflowStatusByIssue]). */
    workflowStatusByIssue: Map<String, String> = emptyMap(),
): List<ReviewRowEntry> {
    val nested = PrStack.nestPrStacks(
        entries,
        { it.branch },
        { it.representative.prBaseBranch },
    )
    // The size of the stack each ROOT starts — the `N pull requests` count.
    val sizeOfRootAt = HashMap<Int, Int>()
    var rootIndex = -1
    nested.forEachIndexed { index, row ->
        if (row.depth == 0) rootIndex = index
        sizeOfRootAt[rootIndex] = (sizeOfRootAt[rootIndex] ?: 0) + 1
    }
    val ancestors = ArrayList<ReviewEntry>()
    var rootBoardId = ""
    rootIndex = -1
    return nested.mapIndexed { index, row ->
        while (ancestors.size > row.depth) ancestors.removeAt(ancestors.size - 1)
        val below = ancestors.lastOrNull()
        if (row.depth == 0) {
            rootIndex = index
            rootBoardId = row.entry.boardId
        }
        ancestors.add(row.entry)
        ReviewRowEntry(
            entry = row.entry,
            depth = row.depth,
            hasChildren = row.hasChildren,
            stackedOn = below?.representative?.identifier,
            rootBoardId = rootBoardId,
            mergeStackIssueId = if (row.depth == 0 && row.hasChildren) {
                row.entry.representative.id
            } else {
                null
            },
            stackSize = sizeOfRootAt[rootIndex] ?: 1,
            workflowStatus = ReviewsMerge.reviewWorkflowStatus(
                row.entry.issues.map { it.id },
                workflowStatusByIssue,
            ),
        )
    }
}

data class ReviewBoardGroup(
    val board: BoardEntity,
    val rows: List<ReviewRowEntry>,
) {
    /** The flat pull requests of this board — counts and callers that ignore nesting. */
    val entries: List<ReviewEntry> get() = rows.map { it.entry }
}

data class ReviewsState(
    val groups: List<ReviewBoardGroup> = emptyList(),
    // EXP-734: issueless runs whose OWN pull request is open — listed under
    // their own header, after the board groups.
    val runs: List<RunReviewEntry> = emptyList(),
    // EXP-1072: workflows whose FINAL pull request is open — listed under
    // their own header, between the board groups and the runs.
    val workflows: List<WorkflowReviewEntry> = emptyList(),
    val loaded: Boolean = false,
) {
    val isEmpty: Boolean get() = groups.isEmpty() && runs.isEmpty() && workflows.isEmpty()
}

@OptIn(ExperimentalCoroutinesApi::class)
@HiltViewModel
class ReviewsViewModel @Inject constructor(
    holder: DatabaseHolder,
    private val auth: AuthRepository,
    private val issuesApi: IssuesApi,
    private val codingSessionsApi: CodingSessionsApi,
    private val workflowsApi: WorkflowsApi,
    selection: TeamSelection,
) : ViewModel() {

    private val dbFlow = accountDatabaseFlow(auth, holder)

    val state: StateFlow<ReviewsState> =
        combine(dbFlow, selection.selectedId) { db, teamId -> db to teamId }
            .flatMapLatest { (db, teamId) ->
                if (db == null || teamId == null) {
                    flowOf(ReviewsState(loaded = true))
                } else {
                    combine(
                        db.issueDao().observeOpenPrsByTeam(teamId),
                        db.boardDao().observeByTeam(teamId),
                        db.codingSessionDao().observeOpenPrRunsByTeam(teamId),
                        db.workflowDao().observeByTeam(teamId),
                        db.workflowNodeDao().observeByTeam(teamId),
                    ) { issues, boards, runs, workflows, nodes ->
                        buildState(issues, boards, runs, workflows, nodes)
                    }
                }
            }
            .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), ReviewsState())

    private fun buildState(
        issues: List<IssueEntity>,
        boards: List<BoardEntity>,
        runs: List<CodingSessionEntity>,
        workflows: List<WorkflowEntity>,
        nodes: List<WorkflowNodeEntity>,
    ): ReviewsState {
        val boardsById = boards.associateBy { it.id }

        // Group by pr_url so a batch PR (N issues, one url) becomes ONE entry;
        // an issue without a url (defensive — the query only selects pr_state
        // 'open', which normally implies a url) keys on its own id so it stays
        // a distinct single-issue row.
        val entries = issues
            .filter { it.boardId in boardsById }
            .groupBy { it.prUrl ?: "issue:${it.id}" }
            .map { (groupKey, rows) ->
                val ordered = rows.sortedByDescending { sortableTimestamp(it.createdAt) }
                val representative = ordered.first()
                ReviewEntry(
                    groupKey = groupKey,
                    prUrl = representative.prUrl,
                    prNumber = representative.prNumber,
                    branch = representative.branch,
                    boardId = representative.boardId,
                    issues = ordered,
                )
            }

        // Newest entry first, then nested by stack (EXP-897): roots keep that
        // order, a stacked pull request follows its foundation. Grouped by the
        // ROOT's board so a stack never splits across two bands, and the
        // boards ordered by sortOrder (name tiebreak) — parity with
        // web/iOS/desktop, which all walk boards in board order.
        // EXP-1094: a node PR of a live workflow merges through the workflow.
        val workflowStatusByIssue = ReviewsMerge.workflowStatusByIssue(
            workflows.map { it.id to it.status },
            nodes.map { it.workflowId to it.coveredIssueIds },
        )
        val rows = buildReviewRows(
            entries.sortedByDescending { sortableTimestamp(it.representative.createdAt) },
            workflowStatusByIssue,
        )
        val groups = rows
            .groupBy { it.rootBoardId }
            .mapNotNull { (boardId, boardRows) ->
                val board = boardsById[boardId] ?: return@mapNotNull null
                ReviewBoardGroup(board = board, rows = boardRows)
            }
            .sortedWith(
                compareBy({ it.board.sortOrder }, { it.board.name.lowercase() })
            )

        return ReviewsState(
            groups = groups,
            runs = buildRunEntries(runs),
            workflows = buildWorkflowEntries(workflows),
            loaded = true,
        )
    }

    /**
     * Squash-merge a RUN's own pull request (EXP-734). No issue is linked, so
     * nothing is completed: the server merges, flips the session row's
     * `pr_state` and (unless the team keeps sessions on merge) ends the run —
     * all of it arriving through Electric, which drops the entry off this
     * list. Shares the merging / mergeErrors maps, keyed by [RunReviewEntry.groupKey].
     */
    fun mergeRun(entry: RunReviewEntry) {
        viewModelScope.launch {
            val accountId = auth.activeAccountId.value ?: return@launch
            val key = entry.groupKey
            _mergeErrors.value = _mergeErrors.value - key
            _merging.value = _merging.value + key
            runCatching { codingSessionsApi.mergePr(accountId, entry.session.id) }
                .onFailure { t ->
                    if (t is CancellationException) throw t
                    _mergeErrors.value = _mergeErrors.value +
                        (key to MergeFailure.from(t, "The pull request could not be merged"))
                }
            _merging.value = _merging.value - key
        }
    }

    /**
     * EXP-1072: squash-merge a workflow's FINAL pull request via
     * `workflows.mergeFinalPr`. The server completes the workflow and moves
     * every landed issue to the team's PR-merge status; the synced row's
     * `final_pr_state` leaving `open` drops the entry. Shares the merging /
     * mergeErrors maps, keyed by [WorkflowReviewEntry.groupKey].
     */
    fun mergeWorkflow(entry: WorkflowReviewEntry) {
        viewModelScope.launch {
            val accountId = auth.activeAccountId.value ?: return@launch
            val key = entry.groupKey
            _mergeErrors.value = _mergeErrors.value - key
            _merging.value = _merging.value + key
            runCatching { workflowsApi.mergeFinalPr(accountId, entry.workflow.id) }
                .onFailure { t ->
                    if (t is CancellationException) throw t
                    _mergeErrors.value = _mergeErrors.value +
                        (key to MergeFailure.from(t, "The pull request could not be merged"))
                }
            _merging.value = _merging.value - key
        }
    }

    /**
     * Squash-merge a review's PR via the GitHub App (EXP-131). Pass the
     * entry's [groupKey] plus the representative issue id — for a batch PR the
     * server resolves it to ALL linked issues and completes them together; the
     * `done` flips arrive via Electric sync, dropping the entry off this list.
     */
    /**
     * EXP-897: merge the whole STACK this row starts, bottom-up. [issueId] is
     * the BOTTOM row's representative issue — the server resolves the chain's
     * top and merges every unmerged member below it, retargeting as it goes.
     * Shares the merging / mergeErrors maps with the single merge.
     */
    fun mergeStack(groupKey: String, issueId: String) {
        viewModelScope.launch {
            val accountId = auth.activeAccountId.value ?: return@launch
            _mergeErrors.value = _mergeErrors.value - groupKey
            _merging.value = _merging.value + groupKey
            runCatching { issuesApi.mergePr(accountId, issueId, mergeStack = true) }
                .onFailure { t ->
                    if (t is CancellationException) throw t
                    _mergeErrors.value = _mergeErrors.value +
                        (groupKey to MergeFailure.from(t, "The stack could not be merged"))
                }
            _merging.value = _merging.value - groupKey
        }
    }

    fun mergePr(groupKey: String, issueId: String) {
        viewModelScope.launch {
            val accountId = auth.activeAccountId.value ?: return@launch
            _mergeErrors.value = _mergeErrors.value - groupKey
            _merging.value = _merging.value + groupKey
            runCatching { issuesApi.mergePr(accountId, issueId) }
                .onFailure { t ->
                    if (t is CancellationException) throw t
                    // Conflicts, branch protection and GitHub App errors are the
                    // COMMON, persistent failures of a squash merge — a silent
                    // drop left the row sitting there unexplained (REV2-50).
                    // Same copy as the issue Changes tab's merge.
                    _mergeErrors.value = _mergeErrors.value +
                        (groupKey to MergeFailure.from(t, "The pull request could not be merged"))
                }
            _merging.value = _merging.value - groupKey
        }
    }

    // Rendered INLINE on the failing row, keyed by its groupKey (EXP-323 — a
    // Scaffold snackbar landed behind the floating bottom nav pill, which is
    // drawn over the whole NavHost, so the reason a merge failed was
    // unreadable). Cleared by the next attempt on that row.
    private val _mergeErrors = MutableStateFlow<Map<String, MergeFailure>>(emptyMap())
    val mergeErrors: StateFlow<Map<String, MergeFailure>> = _mergeErrors

    private val _merging = MutableStateFlow<Set<String>>(emptySet())
    val merging: StateFlow<Set<String>> = _merging
}
